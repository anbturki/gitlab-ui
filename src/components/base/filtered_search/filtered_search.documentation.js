import * as description from './filtered_search.md';
import examples from './examples';

export default {
  followsDesignSystem: true,
  description,
  bootstrapComponent: null,
  examples,
  propsInfo: {
    value: {
      additionalInfo: 'If provided, used as value of filtered search',
    },
    historyItems: {
      type: Array,
      required: false,
      additionalInfo: 'If provided, used as history items for this component',
    },
    availableTokens: {
      type: Array,
      required: true,
      additionalInfo: 'Available tokens',
    },
    placeholder: {
      type: String,
      required: false,
      additionalInfo: 'If provided, used as history items for this component',
    },
    recentSearchesHeader: {
      type: String,
      required: false,
      default: 'Recent searches',
      additionalInfo: 'i18n for recent searches title within history dropdown',
    },
    clearButtonTitle: {
      type: String,
      required: false,
      default: 'Clear',
      additionalInfo: 'i18n for clear button title',
    },
    closeButtonTitle: {
      type: String,
      required: false,
      default: 'Close',
      additionalInfo: 'i18n for close button title within history dropdown',
    },
    clearRecentSearchesText: {
      type: String,
      required: false,
      default: 'Clear recent searches',
      additionalInfo: 'i18n for recent searches clear text',
    },
  },
  events: [
    {
      event: 'submit',
      args: [
        {
          arg: 'tokens',
          description: '(Array)',
        },
      ],
      description: 'Emitted when search is submitted',
    },
  ],
  slots: [
    {
      name: 'history-item',
      description:
        'Slot to customize history item in history dropdown. Used only if using history items',
    },
  ],
};
