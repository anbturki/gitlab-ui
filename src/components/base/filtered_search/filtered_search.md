# Filtered Search

The filtered search component is responsible for managing search with possible filters.

## Usage

Each filter option (named token) requires a separate Vue component. `GlFilteredSearchToken` is an example of such a token.

Prepare array of available token configurations with the following fields:

- `type`: unique identifier of token type
- `title`: human-readable title of the token
- `icon`: token icon
- `dataType`: (optional) identifier of type (for example "user") for this filter. Tokens
  of the same type could be switched without loosing their values
- `unique`: (optional) indicate this token could appear only once in the filter
- `disabled`: (optional) indicate this token should be hidden from the dropdown
- any additional fields required to configure your component

Each token for filtered search is a Vue component with the following props:

- `active`: indicates if the token is currently active. It's the token's responsibility
  to render proper control for editing (for example input).
- `current-value`: current tokens of the filtered search.
- `index`: current token position in the filtered search.
- `config`: additional configuration, supplied in filtered search config for this token.

The token should emit the following events:

- `activate`: when the token requests activation (for example, when being clicked).
- `deactivate`: when token requests deactivation (for example due to losing blur on input).
- `destroy`: when token requests self-destruction (for instance for clicking "X" sign).
- `replace`: token requests its replacement with another token.
- `split`: token requests adding string values after the current token.
- `complete`: token indicates its editing is completed.

## Examples

Define a list of available tokens:

```js
const availableTokens = [
  { type: 'static', icon: 'label', title: 'static:token', token: staticToken },
  { type: 'dynamic', icon: 'rocket', title: 'dynamic:~token', token: dynamicToken },
];
```

Pass the list of tokens to the search component. Optionally, you can use `v-model` to receive realtime updates:

```js
 <gl-filtered-search :available-tokens="tokens" v-model="value" />
```
