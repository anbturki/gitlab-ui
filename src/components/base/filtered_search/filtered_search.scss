.gl-filtered-search-scrollable {
  @include gl-pl-2;
  @include gl-display-flex;
  @include gl-flex-grow-1;
  @include gl-flex-basis-0;
  @include gl-overflow-y-auto;
  @include gl-align-items-center;
  @include gl-inset-border-1-gray-100;
}

.gl-filtered-search-item {
  @include gl-px-1;
}

.gl-filtered-search-last-item {
  .gl-filtered-search-term-value {
    @include gl-h-full;
    @include gl-w-full;
    @include gl-align-items-center;
  }
  @include gl-flex-grow-1;
  @include gl-p-0;
  @include gl-pl-1;
  @include gl-h-6;
  min-width: $gl-spacing-scale-12;
}
