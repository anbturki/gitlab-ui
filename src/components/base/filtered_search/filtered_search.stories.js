import { withKnobs } from '@storybook/addon-knobs';
import { documentedStoriesOf } from '../../../../documentation/documented_stories';
import readme from './filtered_search.md';
import { GlFilteredSearch, GlFilteredSearchToken } from '../../../../index';

const fakeUsers = [
  { id: 1, name: 'User Alpha', username: 'alpha' },
  { id: 2, name: 'User Beta', username: 'beta' },
  { id: 3, name: 'User Gamma', username: 'gamma' },
  { id: 4, name: 'User Delta', username: 'delta' },
  { id: 5, name: 'User Epsilon', username: 'epsilon' },
];

const fakeMilestones = [
  { id: 1, title: '12.7', name: '%12.7' },
  { id: 2, title: '12.8', name: '%12.8' },
  { id: 3, title: '12.9', name: '%12.9' },
  { id: 4, title: '12.10', name: '%12.10' },
  { id: 5, title: 'Backlog', name: 'Backlog' },
];

const UserToken = {
  props: ['value', 'active'],
  inheritAttrs: false,
  data() {
    return {
      loadingView: false,
      loadingSuggestions: false,
      users: [],
      activeUser: null,
    };
  },
  methods: {
    loadView() {
      this.loadingView = true;
      setTimeout(() => {
        this.loadingView = false;
        this.activeUser = fakeUsers.find(u => u.username === this.value.data);
      }, 1000);
    },
    loadSuggestions() {
      this.loadingSuggestions = true;
      setTimeout(() => {
        this.loadingSuggestions = false;
        this.users = fakeUsers;
      }, 2000);
    },
  },
  watch: {
    // eslint-disable-next-line func-names
    'value.data': function() {
      if (this.active) {
        this.loadSuggestions();
      }
    },
    active: {
      immediate: true,
      handler(newValue) {
        if (!newValue) {
          this.loadView();
        } else {
          this.loadSuggestions();
        }
      },
    },
  },
  template: `
    <gl-filtered-search-token
      v-bind="{ ...this.$props, ...this.$attrs }"
      v-on="$listeners"
    >
      <template #view="{ inputValue }">
        <gl-loading-icon size="sm" v-if="loadingView" class="gl-mr-2" />
        <gl-avatar :size="16" :entity-name="inputValue" shape="circle" class="gl-mr-2" v-else />
        {{ activeUser ? activeUser.name : inputValue }}
      </template>
      <template #suggestions>
        <template v-if="loadingSuggestions">
          <gl-loading-icon />
        </template>
        <template v-else>
        <gl-filtered-search-suggestion :key="user.id" v-for="user in users" :value="user.username">
          <div class="gl-display-flex">
            <gl-avatar :size="32" :entity-name="user.username" />
            <div>
              <p class="gl-m-0">{{ user.name }}</p>
              <p class="gl-m-0">@{{ user.username }}</p>
            </div>
          </div>
        </gl-filtered-search-suggestion>
        </template>
      </template>
    </gl-filtered-search-token>
  `,
};

const MilestoneToken = {
  props: ['value', 'active'],
  inheritAttrs: false,
  data() {
    return {
      loadingView: false,
      loadingSuggestions: false,
      milestones: [],
    };
  },
  methods: {
    loadSuggestions() {
      this.loadingSuggestions = true;
      setTimeout(() => {
        this.loadingSuggestions = false;
        this.milestones = fakeMilestones;
      }, 2000);
    },
  },
  watch: {
    // eslint-disable-next-line func-names
    'value.data': function() {
      if (this.active) {
        this.loadSuggestions();
      }
    },
    active: {
      immediate: true,
      handler(newValue) {
        if (newValue) {
          this.loadSuggestions();
        }
      },
    },
  },
  template: `
    <gl-filtered-search-token
      v-bind="{ ...this.$props, ...this.$attrs }"
      v-on="$listeners"
    >
      <template #suggestions>
        <gl-filtered-search-suggestion value="None">None</gl-filtered-search-suggestion>
        <gl-filtered-search-suggestion value="Any">Any</gl-filtered-search-suggestion>
        <gl-filtered-search-suggestion value="Upcoming">Upcoming</gl-filtered-search-suggestion>
        <gl-filtered-search-suggestion value="Started">Started</gl-filtered-search-suggestion>
        <gl-dropdown-divider v-if="loadingSuggestions || milestones.length" />
        <template v-if="loadingSuggestions">
          <gl-loading-icon />
        </template>
        <template v-else>
        <gl-filtered-search-suggestion :key="milestone.id" v-for="milestone in milestones" :value="milestone.name">
        {{ milestone.title }}
        </gl-filtered-search-suggestion>
        </template>
      </template>
    </gl-filtered-search-token>
  `,
};

const tokens = [
  {
    type: 'author',
    icon: 'pencil',
    title: 'Author',
    dataType: 'user',
    unique: true,
    token: UserToken,
  },
  { type: 'user', icon: 'user', title: 'Assignee', dataType: 'user', token: UserToken },
  { type: 'milestone', icon: 'clock', title: 'Milestone', unique: true, token: MilestoneToken },
  { type: 'weight', icon: 'weight', title: 'Weight', unique: true, token: GlFilteredSearchToken },
  {
    type: 'confidential',
    icon: 'eye-slash',
    title: 'Confidential',
    unique: true,
    token: GlFilteredSearchToken,
    options: [
      { icon: 'eye-slash', value: true, title: 'Yes' },
      { icon: 'eye', value: false, title: 'No' },
    ],
  },
];

const components = {
  GlFilteredSearch,
};

documentedStoriesOf('base|filtered-search', readme)
  .addDecorator(withKnobs)
  .add('default', () => ({
    data() {
      return {
        tokens,
        value: [{ type: 'user', value: { data: 'beta', operator: '=' } }, 'raw text'],
      };
    },
    components,
    computed: {
      formattedValue() {
        return JSON.stringify(this.value);
      },
    },
    template: `
      <div>
        <pre>{{ formattedValue }}</pre>
        <gl-filtered-search :available-tokens="tokens" v-model="value" />
      </div>
    `,
  }));
