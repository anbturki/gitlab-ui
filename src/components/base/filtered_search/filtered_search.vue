<script>
import Vue from 'vue';
import PortalVue from 'portal-vue';
import GlSearchBoxByClick from '../search_box_by_click/search_box_by_click.vue';
import GlFilteredSearchTerm from './filtered_search_term.vue';
import GlIcon from '../icon/icon.vue';
import {
  isEmptyTerm,
  TERM_TOKEN_TYPE,
  normalizeTokens,
  denormalizeTokens,
  needDenormalization,
} from './filtered_search_utils';
import GlTooltip from '../../../directives/tooltip';

Vue.use(PortalVue);

let portalUuid = 0;

function createTerm(data = '') {
  return {
    type: TERM_TOKEN_TYPE,
    value: { data },
  };
}

function initialState() {
  return [createTerm()];
}

export default {
  components: {
    GlSearchBoxByClick,
    GlIcon,
  },
  directives: { GlTooltip },
  inheritAttrs: false,
  props: {
    value: {
      required: false,
      type: Array,
      default: () => [],
    },
    availableTokens: {
      type: Array,
      required: true,
    },
    placeholder: {
      type: String,
      required: false,
      default: 'Search',
    },
    clearButtonTitle: {
      type: String,
      default: 'Clear',
    },
  },
  data() {
    return {
      tokens: initialState(),
      activeTokenIdx: null,
      suggestionsStyle: {},
    };
  },
  provide() {
    portalUuid += 1;
    this.portalName = `filters_portal_${portalUuid}`;

    return {
      portalName: this.portalName,
      alignSuggestions: ref => this.alignSuggestions(ref),
    };
  },
  computed: {
    activeToken() {
      return this.tokens[this.activeTokenIdx];
    },
    lastTokenIdx() {
      return this.tokens.length - 1;
    },
    isLastTokenActive() {
      return this.activeTokenIdx === this.lastTokenIdx;
    },
    hasValue() {
      return this.tokens.length > 1 || this.tokens[0].value.data !== '';
    },
    termPlaceholder() {
      return this.hasValue ? null : this.placeholder;
    },

    currentAvailableTokens() {
      return this.availableTokens.filter(token => {
        if (token.disabled) {
          return false;
        }

        if (token.unique) {
          return !this.tokens.find(t => t.type === token.type);
        }

        return true;
      });
    },
  },
  watch: {
    value: {
      immediate: true,
      handler(newValue) {
        this.tokens = needDenormalization(newValue) ? denormalizeTokens(newValue) : newValue;
        if (this.tokens.length === 0 || this.tokens[this.lastTokenIdx].type !== TERM_TOKEN_TYPE) {
          this.tokens.push(createTerm());
        }
      },
    },
    tokens: {
      handler() {
        this.$emit('input', this.tokens);
      },
      deep: true,
      immediate: true,
    },
  },

  methods: {
    isLastToken(idx) {
      return !this.activeTokenIdx && idx === this.lastTokenIdx;
    },

    getTokenEntry(type) {
      return this.availableTokens.find(t => t.type === type);
    },

    getTokenComponent(type) {
      return this.getTokenEntry(type)?.token || GlFilteredSearchTerm;
    },

    tokenConfig(type) {
      const cmp = this.getTokenEntry(type);
      if (!cmp) {
        return null;
      }

      const { token, ...config } = cmp;

      return config;
    },

    activate(token) {
      this.activeTokenIdx = token;
    },

    alignSuggestions(ref) {
      const offsetRef = ref.getBoundingClientRect().left;
      const offsetMenu = this.$el.getBoundingClientRect().left;
      const transform = `translateX(${Math.floor(offsetRef - offsetMenu)}px)`;
      this.suggestionsStyle = { transform };
    },

    deactivate(tokenIdx, type) {
      if (tokenIdx === -1 || this.activeTokenIdx !== tokenIdx || this.activeToken.type !== type) {
        return;
      }

      if (
        !this.isLastTokenActive &&
        this.activeToken.type === TERM_TOKEN_TYPE &&
        this.activeToken.value.data.trim() === ''
      ) {
        this.tokens.splice(tokenIdx, 1);
      }

      this.activeTokenIdx = null;
    },

    destroyToken(idx) {
      if (this.tokens.length === 1) {
        return;
      }

      this.tokens.splice(idx, 1);
      if (idx !== 0) {
        this.activeTokenIdx = idx - 1;
      }
    },

    replaceToken(idx, token) {
      this.$set(this.tokens, idx, { ...token, value: { data: '', ...token.value } });
      this.activeTokenIdx = idx;
    },

    createTokens(idx, newStrings = []) {
      if (
        this.activeTokenIdx !== this.lastTokenIdx &&
        newStrings.length === 1 &&
        newStrings[0] === ''
      ) {
        this.activeTokenIdx = this.lastTokenIdx;
        return;
      }

      this.tokens.splice(
        idx + 1,
        0,
        ...newStrings.map(data => ({
          type: TERM_TOKEN_TYPE,
          value: { data },
        }))
      );

      this.activeTokenIdx = idx + newStrings.length;
    },

    completeToken() {
      if (this.activeTokenIdx === this.lastTokenIdx - 1) {
        this.activeTokenIdx = this.lastTokenIdx;
      } else {
        this.activeTokenIdx = null;
      }
    },

    submit() {
      this.$emit('submit', normalizeTokens(this.tokens));
    },

    clearInput() {
      this.tokens = initialState();
    },
  },
};
</script>

<template>
  <gl-search-box-by-click v-bind="$attrs" @submit="submit">
    <template #history-item="slotScope">
      <slot name="history-item" v-bind="slotScope"></slot>
    </template>
    <template #input>
      <div class="gl-filtered-search-scrollable">
        <template v-for="(token, idx) in tokens">
          <component
            :is="getTokenComponent(token.type)"
            ref="tokens"
            :key="`${token.type}-${idx}`"
            v-model="token.value"
            v-bind="tokenConfig(token.type)"
            :active="activeTokenIdx === idx"
            :available-tokens="currentAvailableTokens"
            :current-value="tokens"
            :index="idx"
            :placeholder="termPlaceholder"
            class="gl-filtered-search-item"
            :class="{
              'gl-filtered-search-last-item': isLastToken(idx),
            }"
            @activate="activate(idx)"
            @deactivate="deactivate(idx, token.type)"
            @destroy="destroyToken(idx)"
            @replace="replaceToken(idx, $event)"
            @complete="completeToken"
            @submit="submit"
            @split="createTokens(idx, $event)"
          />
        </template>
      </div>
      <button
        v-if="hasValue"
        v-gl-tooltip.hover
        :title="clearButtonTitle"
        class="gl-search-box-by-click-icon-button gl-search-box-by-click-clear-button"
        name="clear"
        @click="clearInput"
      >
        <gl-icon name="clear" />
      </button>
      <portal-target
        ref="menu"
        :key="activeTokenIdx"
        :name="portalName"
        slim
        :style="suggestionsStyle"
      />
    </template>
  </gl-search-box-by-click>
</template>
