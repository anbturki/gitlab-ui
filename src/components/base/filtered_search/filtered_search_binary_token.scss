.gl-filtered-search-binary-token {
  @include gl-display-inline-flex;
  @include gl-white-space-nowrap;
}

.gl-filtered-search-binary-token-type {
  @include gl-bg-gray-100;
  @include gl-rounded-top-right-none;
  @include gl-rounded-bottom-right-none;
}

.gl-filtered-search-binary-token-value {
  @include gl-bg-gray-200;
  @include gl-rounded-top-left-none;
  @include gl-rounded-bottom-left-none;
}

.gl-filtered-search-binary-token-input {
  @include gl-p-0;
  @include gl-pl-2;
  @include gl-border-none;
  @include gl-outline-none;
  min-width: $gl-dropdown-width;
}
