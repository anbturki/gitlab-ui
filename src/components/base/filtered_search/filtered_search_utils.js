export const TERM_TOKEN_TYPE = 'filtered-search-term';

export function isEmptyTerm(token) {
  return token.type === TERM_TOKEN_TYPE && token.value.trim() === '';
}

export function normalizeTokens(tokens) {
  const result = [];
  tokens.forEach(token => {
    if (isEmptyTerm(token)) {
      return;
    }

    if (token.type !== TERM_TOKEN_TYPE) {
      result.push({ ...token });
    } else if (result.length > 0 && typeof result[result.length - 1] === 'string') {
      result[result.length - 1] += ` ${token.value}`;
    } else {
      result.push(token.value);
    }
  });
  return result;
}

export function needDenormalization(tokens) {
  return tokens.some(t => typeof t === 'string');
}

export function denormalizeTokens(tokens) {
  const result = [];
  tokens.forEach(t => {
    if (typeof t === 'string') {
      const stringTokens = t.split(' ').filter(Boolean);
      stringTokens.forEach(strToken => result.push({ type: TERM_TOKEN_TYPE, value: strToken }));
    } else {
      result.push(t);
    }
  });
  return result;
}
