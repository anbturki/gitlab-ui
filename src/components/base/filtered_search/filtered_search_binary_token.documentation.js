import * as description from './filtered_search_binary_token.md';

export default {
  description,
  bootstrapComponent: null,
  propsInfo: {
    title: {
      additionalInfo: 'Token title',
    },
    active: {
      additionalInfo: 'If this token is currently active',
    },
    value: {
      additionalInfo: 'Current value',
    },
  },
  slots: [
    { name: 'view', description: 'Template for token value in inactive state' },
    { name: 'suggestions', description: 'Slot for rendering autocomplete suggestions' },
  ],
};
