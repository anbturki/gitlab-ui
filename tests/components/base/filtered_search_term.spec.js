import { shallowMount, createLocalVue } from '@vue/test-utils';
import PortalVue from 'portal-vue';
import FilteredSearchTerm from '../../../src/components/base/filtered_search/filtered_search_term.vue';
import GlFilteredSearchSuggestion from '../../../src/components/base/filtered_search/filtered_search_suggestion.vue';
import { TERM_TOKEN_TYPE } from '../../../src/components/base/filtered_search/filtered_search_utils';

const localVue = createLocalVue();
localVue.use(PortalVue);

const availableTokens = [
  { type: 'foo', hint: 'test1-foo', token: 'stub', icon: 'eye' },
  { type: 'foo', hint: 'test2-bar', token: 'stub', icon: 'eye' },
  { type: 'foo', hint: 'test1-baz', token: 'stub', icon: 'eye' },
];

describe('Filtered search term', () => {
  let wrapper;

  beforeAll(() => {
    if (!HTMLElement.prototype.scrollIntoView) {
      HTMLElement.prototype.scrollIntoView = jest.fn();
    }
  });

  afterAll(() => {
    if (HTMLElement.prototype.scrollIntoView.mock) {
      delete HTMLElement.prototype.scrollIntoView;
    }
  });

  const defaultProps = {
    availableTokens: [],
  };

  let alignSuggestionsMock;
  let suggestionsMock;

  const createComponent = props => {
    alignSuggestionsMock = jest.fn();
    suggestionsMock = {
      methods: { nextItem: jest.fn(), prevItem: jest.fn(), getValue: jest.fn() },
      template: `<div><slot></slot></div>`,
    };

    wrapper = shallowMount(FilteredSearchTerm, {
      propsData: { ...defaultProps, ...props },
      provide: {
        alignSuggestions: alignSuggestionsMock,
        portalName: 'stub',
      },
      localVue,
      stubs: {
        GlFilteredSearchSuggestionList: suggestionsMock,
      },
    });
  };

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  it('renders value in inactive mode', () => {
    createComponent({ value: { data: 'test-value' } });
    expect(wrapper.text()).toContain('test-value');
  });

  it('renders input with value in active mode', () => {
    createComponent({ value: { data: 'test-value' }, active: true });
    expect(wrapper.find('input').element.value).toBe('test-value');
  });

  it('renders input with placeholder if placeholder prop is provided', () => {
    createComponent({ placeholder: 'placeholder-stub' });
    expect(wrapper.find('input').attributes('placeholder')).toBe('placeholder-stub');
  });

  it('focuses and scrolls to input on activation', () => {
    createComponent({ placeholder: 'placeholder-stub', availableTokens });
    wrapper.setProps({ active: true });
    jest.spyOn(HTMLInputElement.prototype, 'focus');
    jest.spyOn(HTMLInputElement.prototype, 'scrollIntoView');
    return wrapper.vm.$nextTick().then(() => {
      expect(HTMLInputElement.prototype.focus).toHaveBeenCalled();
      expect(HTMLInputElement.prototype.scrollIntoView).toHaveBeenCalled();
    });
  });

  it('aligns suggestions on activation', () => {
    createComponent({ availableTokens });
    wrapper.setProps({ active: true });
    return wrapper.vm.$nextTick().then(() => {
      expect(alignSuggestionsMock).toHaveBeenCalledWith(wrapper.find('input').element);
    });
  });

  it('emits deactivate event when blurred', () => {
    createComponent({ placeholder: 'placeholder-stub', active: true });
    wrapper.find('input').trigger('blur');
    return wrapper.vm.$nextTick().then(() => {
      expect(wrapper.emitted().deactivate).toBeDefined();
    });
  });

  it('renders suggestions when token is active', () => {
    createComponent({ availableTokens, active: true });
    expect(wrapper.find(suggestionsMock).exists()).toBe(true);
    expect(wrapper.findAll(GlFilteredSearchSuggestion).length).toBe(availableTokens.length);
  });

  it('filters suggestions by input', () => {
    createComponent({ availableTokens, active: true, value: { data: 'test1' } });
    wrapper.find('input').setValue('test1');
    return wrapper.vm.$nextTick().then(() => {
      expect(wrapper.findAll(GlFilteredSearchSuggestion).length).toBe(2);
    });
  });

  it('emits submit event if no suggestion is selected and Enter is pressed', () => {
    createComponent({ availableTokens, active: true });
    wrapper.find('input').trigger('keydown', { key: 'Enter' });
    return wrapper.vm.$nextTick().then(() => {
      expect(wrapper.emitted().submit.length).toBe(1);
    });
  });

  it('emits submit event if no suggestions are available and Enter is pressed', () => {
    createComponent({ availableTokens, active: true, value: { data: 'other' } });
    wrapper.find('input').trigger('keydown', { key: 'Enter' });
    return wrapper.vm.$nextTick().then(() => {
      expect(wrapper.emitted().submit.length).toBe(1);
    });
  });

  it('emits replace event if suggestion is selected and Enter is pressed', () => {
    createComponent({ availableTokens, active: true });
    suggestionsMock.methods.getValue.mockReturnValue('token-type');
    wrapper.find('input').trigger('keydown', { key: 'Enter' });
    return wrapper.vm.$nextTick().then(() => {
      expect(wrapper.emitted().replace.length).toBe(1);
      expect(wrapper.emitted().replace[0]).toEqual([{ type: 'token-type' }]);
    });
  });

  it('emits destroy event if value is empty and Backspace is pressed', () => {
    createComponent({ availableTokens, active: true });
    wrapper.find('input').trigger('keydown', { key: 'Backspace' });
    return wrapper.vm.$nextTick().then(() => {
      expect(wrapper.emitted().destroy.length).toBe(1);
    });
  });

  it('emits create when value is changed to ending with single space', () => {
    createComponent({ availableTokens, active: true });
    wrapper.setProps({ value: { data: 'test ' } });
    return wrapper.vm.$nextTick().then(() => {
      expect(wrapper.emitted().create.length).toBe(1);
      expect(wrapper.emitted().create[0]).toEqual([
        [{ type: TERM_TOKEN_TYPE, value: { data: '' } }],
      ]);
    });
  });

  it('does not emit create when value is changed to contain multiple spaces and has unclosed quotes', () => {
    createComponent({ availableTokens, active: true });
    wrapper.setProps({ value: { data: 'foo "bar baz' } });
    return wrapper.vm.$nextTick().then(() => {
      expect(wrapper.emitted().create).toBeUndefined();
    });
  });

  it('emits create and truncates current token when value is changed to contain multiple spaces and has no quotes', () => {
    createComponent({ availableTokens, active: true });
    wrapper.setProps({ value: { data: 'foo bar baz' } });
    return wrapper.vm.$nextTick().then(() => {
      expect(wrapper.emitted().create[0]).toEqual([
        [
          { type: TERM_TOKEN_TYPE, value: { data: 'bar' } },
          { type: TERM_TOKEN_TYPE, value: { data: 'baz' } },
        ],
      ]);
      const lastEmit = wrapper.emitted().input[wrapper.emitted().input.length - 1];
      expect(lastEmit).toEqual([{ data: 'foo' }]);
    });
  });

  it('emits create and truncates current token when value is changed to contain multiple spaces and has paired quotes', () => {
    createComponent({ availableTokens, active: true });
    wrapper.setProps({ value: { data: '"foo bar" baz' } });
    return wrapper.vm.$nextTick().then(() => {
      expect(wrapper.emitted().create[0]).toEqual([
        [{ type: TERM_TOKEN_TYPE, value: { data: 'baz' } }],
      ]);
      const lastEmit = wrapper.emitted().input[wrapper.emitted().input.length - 1];
      expect(lastEmit).toEqual([{ data: '"foo bar"' }]);
    });
  });

  it('does not emit create event if there are no spaces in input', () => {
    createComponent({ availableTokens, active: true });
    wrapper.setProps({ value: { data: 'foo' } });
    return wrapper.vm.$nextTick().then(() => {
      expect(wrapper.emitted().create).toBeUndefined();
    });
  });

  it('selects next suggestion if down arrow is pressed', () => {
    createComponent({ availableTokens, active: true });
    wrapper.find('input').trigger('keydown', { key: 'ArrowDown' });
    expect(suggestionsMock.methods.nextItem).toHaveBeenCalled();
  });

  it('selects previous suggestion if up arrow is pressed', () => {
    createComponent({ availableTokens, active: true });
    wrapper.find('input').trigger('keydown', { key: 'ArrowUp' });
    expect(suggestionsMock.methods.prevItem).toHaveBeenCalled();
  });
});
