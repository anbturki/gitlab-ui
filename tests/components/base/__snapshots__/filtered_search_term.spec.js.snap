// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filtered search term renders input with value in active mode 1`] = `
<div class="gl-h-auto">
  <div active="true" value="test-value" class="">test-value</div>
</div>
`;

exports[`Filtered search term renders value in inactive mode 1`] = `
<div class="gl-h-auto">
  <div value="test-value" class="">test-value</div>
</div>
`;
